llvm/lib/Analysis/TargetLibraryInfo.cpp

```diff
@@ -1465,13 +1465,19 @@ unsigned TargetLibraryInfoImpl::getSizeTSize(const Module &M) const {
 
   // Historically LLVM assume that size_t has same size as intptr_t (hence
   // deriving the size from sizeof(int*) in address space zero). This should
-  // work for most targets. For future consideration: DataLayout also implement
-  // getIndexSizeInBits which might map better to size_t compared to
-  // getPointerSizeInBits. Hard coding address space zero here might be
-  // unfortunate as well. Maybe getDefaultGlobalsAddressSpace() or
-  // getAllocaAddrSpace() is better.
+  // work for most targets. For future consideration: Hard coding address space
+  // zero here might be unfortunate. Maybe getMaxIndexSizeInBits() is better.
```

Comment:
That would be better

Reply:
Probably not a big difference, but...
`size_t` is a C language concept not directly related to data layout, which is more of a machine concept.
The size of `size_t` can as well be less than whatever `get*IndexSizeInBits` returns.
Would it make more sense to handle `size_t` similar to `int`? (See `getIntSize`, `setIntSize`.)

Reply:
You mean rather than determine this from the data layout compute it from the triple instead? I think that would be fine too.

Reply:
> You mean rather than determine this from the data layout compute it from the triple instead?

Yes.

I can also see the size of `size_t` and sizes of other types used by library functions stored in metadata (as it is for `wchar_t`), but it probably doesn't worth the effort.

Reply:
I don't think there's a huge difference in the approaches here. `getSizeTSize` already takes a `Module` argument specifically to enable target-specific overriding of the default behavior, but as of yet we don't have a target that can't be handle purely by the generic logic.

I'll address @arsenm's feedback and update the comment in this file to make it clear that targets may add overrides in this function.

Reply:
It turns out that using `getMaxIndexSizeInBits` is not a better default, as it computes the wrong size for a number of currently supported targets, including i386, x86-64 x32, and aaarch64 x32 (quite possibly others, but those are the ones I detected quickly).

I have updated the comment to reflect this reality.

@s-barannikov In the process I did experiment with per-target overrides in `getSizeTSize` and it worked without a hitch. If there were only a single configuration that needed overriding that would have been the way to go, but as it turned out the default behavior was clearly suboptimal.

Reply:
I think the main advantage of using a triple-based implementation vs the data layout heuristic is that we could save the Module argument on the method...

Reply:
I tested initializing it in `initializeBase`, as we do for `IntSize`. The problem with doing it there is that we lack access to the `DataLayout`, so the best we can do would be to default to `Triple::getArchPointerBitWidth()`, which wouldn't represent an improvement on the status quo.

Reply:
Well, yes, quering Triple instead of DataLayout was my point...

Reply:
That defeats the purpose of this PR.

Reply:
Consider this resolved

Reply:
Couldn't we introduce getArchIndexBitWidth() which is basically the pointer width except on cheri targets (which should be distinguishable by target triple, right?)